<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="QurvE">
<title>Quantitiative Fluorescence Curve Evaluation with Package `QurvE` • QurvE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quantitiative Fluorescence Curve Evaluation with Package `QurvE`">
<meta property="og:description" content="QurvE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">QurvE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/shiny_app_manual.html">QurvE User Manual v.1.0</a>
    <a class="dropdown-item" href="../articles/vignette_fluorescence.html">Quantitiative Fluorescence Curve Evaluation with Package `QurvE`</a>
    <a class="dropdown-item" href="../articles/vignette_growth.html">Quantitiative Growth Curve Evaluation with Package `QurvE`</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/NicWir/QurvE/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Quantitiative Fluorescence Curve Evaluation with Package `QurvE`</h1>
                        <h4 data-toc-skip class="author">Nicolas T.
Wirth</h4>
            
            <h4 data-toc-skip class="date">2022-12-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/NicWir/QurvE/blob/HEAD/vignettes/vignette_fluorescence.Rmd" class="external-link"><code>vignettes/vignette_fluorescence.Rmd</code></a></small>
      <div class="d-none name"><code>vignette_fluorescence.Rmd</code></div>
    </div>

    
    
<p><strong>Please note:</strong> This vignette will be updated from time
to time when new features are implemented. Please find the most recent
version at the <a href="https://github.com/NicWir/QurvE" class="external-link">QurvE GitHub
repository</a>.</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>For a general introduction to the package, see the vignette
<em>Quantitiative Growth Curve Evaluation with Package
<code>QurvE</code></em>. This vignette will show you how to use
<code>QurvE</code> to analyze datasets with fluorescence measurements
taken over time, as well as how to extract the most important
results.</p>
<p>Fluorescent reporter proteins are widely used to study the mechanisms
of gene regulation or to develop biosensors that allow the detection of
chemical compounds and provide insights into the intracellular
environment. A workflow for analyzing fluorescence data works
analogously to analyzing growth data by following the scheme <em>read
data</em> and <em>execute workflow</em>. The package allows users to
select either time or growth data (e.g. from the simultaneous
measurement of cell density and fluorescence intensity in plate reader
experiments) as the independent variable. Additionally, biosensors can
be characterized via dose-response analysis.</p>
</div>
<div class="section level2">
<h2 id="fluorescence-profiling-methods">Fluorescence profiling methods<a class="anchor" aria-label="anchor" href="#fluorescence-profiling-methods"></a>
</h2>
<p><code>QurvE</code> offers the same curve evaluation methods for
fluorescence data as for growth data, with the exception of fitting
growth models. The calculation algorithms for <em>linear regression</em>
and <em>nonparametric fits</em> (smoothing splines) as well as the
default parameters have been empirically adjusted to account for the
differences in curve shapes commonly seen with fluorescence data.</p>
</div>
<div class="section level2">
<h2 id="dose-response-analysis-methods">Dose-response analysis methods<a class="anchor" aria-label="anchor" href="#dose-response-analysis-methods"></a>
</h2>
<p>Dose response analysis is a useful method for evaluating the
performance of a biosensor. Biosensors are typically designed to respond
to specific chemical compounds, and the strength of the response often
depends on the concentration of the target compound. Dose-response
analysis can be used to determine the concentration of a target compound
that elicits a half-maximal response in the biosensor (variants), the
half-maximal effective concentration (EC<sub>50</sub>). This allows
evaluation of the <em>sensitivity</em> and <em>specificity</em> of a
biosensor and can be used to optimize the design for a particular
application. In addition to evaluating the performance of biosensors,
dose-response analysis can also be used to study the mechanisms of gene
regulation in biological systems. By measuring the response of a
biological system to different concentrations of a chemical compound,
researchers can gain insights into the genes and pathways involved in
the response, and can identify potential targets for drug discovery.
<code>QurvE</code> provides two methods to perform dose-response
analyses on fluorescence data:</p>
<ol style="list-style-type: decimal">
<li><p>Perform a smooth spline fit on response vs. concentration data
and extract the <em>EC<sub>50</sub></em> as the concentration at the
midpoint between the largest and smallest response value.</p></li>
<li><p>Apply a biosensor response model to response vs. concentration
data <span class="citation">(<a href="#ref-meyer2019" role="doc-biblioref">Meyer <em>et al.</em>, 2019</a>)</span>.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="run-a-complete-fluorescence-analysis-workflow">Run a complete fluorescence analysis workflow<a class="anchor" aria-label="anchor" href="#run-a-complete-fluorescence-analysis-workflow"></a>
</h2>
<p>Load the package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/NicWir/QurvE" class="external-link">QurvE</a></span><span class="op">)</span></span></code></pre></div>
<p>Next, load your experimental data. In this example, the dataset being
used is from a preliminary characterization of different versions of the
SEVA (Standard European Vector Architecture) plasmid pSEVA634, as
described in <span class="citation">(<a href="#ref-nikel2022" role="doc-biblioref">Nikel <em>et al.</em>, 2022</a>)</span>. The data
contains both growth an fluorescence measurements that have been
converted into the custom <code>QurvE</code> data format and are located
in different work sheets of the same XLSX file:</p>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_data.html">read_data</a></span><span class="op">(</span>data.growth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"lac_promoters.xlsx"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"QurvE"</span><span class="op">)</span>, sheet.growth <span class="op">=</span> <span class="fl">1</span>, data.fl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"lac_promoters.xlsx"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"QurvE"</span><span class="op">)</span>, sheet.fl <span class="op">=</span> <span class="fl">2</span>, fl.normtype <span class="op">=</span> <span class="st">"growth"</span><span class="op">)</span>  <span class="co"># normalize fluorescence to growth data</span></span></code></pre></div>
<p>The two functions <code><a href="../reference/read_data.html">read_data()</a></code> or
<code><a href="../reference/parse_data.html">parse_data()</a></code> come with more arguments to give the user more
control when loading fluorescence data. As for growth data, arguments
<span style="color: red;"><code>data.fl</code></span> (file path), <span style="color: red;"><code>csvsep.fl</code></span> (separator symbol in
CSV file), <span style="color: red;"><code>dec.fl</code></span> (decimal
separator), and <span style="color: red;"><code>sheet.fl</code></span>
(Excel file worksheet number or “name”) provide details on how where an
how the data is stored. <span style="color: red;"><code>calib.fl</code></span> allows defining an
equation with which to transform fluorescence values.</p>
<p>Similarly, the functions accept the arguments <span style="color: red;"><code>data.fl2</code></span>, <span style="color: red;"><code>csvsep.fl2</code></span>, <span style="color: red;"><code>dec.fl2</code></span>, <span style="color: red;"><code>sheet.fl2</code></span>, and <span style="color: red;"><code>calib.fl2</code></span> to load data from a
second fluorescence channel. This second fluorescence is currently only
used to normalize the first fluorescence values, as applied in …ADD
CITATION…</p>
<p>Normalization of fluorescence, if any, can be controlled via <span style="color: red;"><code>fl.normtype</code></span> to be performed by
either dividing by growth values (<span style="color: red;"><code>fl.normtype = 'growth'</code></span>) or
fluorescence 2 <span style="color: red;"><code>fl.normtype = 'fl2'</code></span>.</p>
<p>We can inspect the structure of the <code>input</code> object of
class <code>grodata</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">input</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Length Class      Mode   </span></span>
<span><span class="co">#&gt; time              1785   -none-     numeric</span></span>
<span><span class="co">#&gt; growth              88   data.frame list   </span></span>
<span><span class="co">#&gt; fluorescence        88   data.frame list   </span></span>
<span><span class="co">#&gt; norm.fluorescence   88   data.frame list   </span></span>
<span><span class="co">#&gt; expdesign            4   data.frame list</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-raw-data">Plot raw data<a class="anchor" aria-label="anchor" href="#plot-raw-data"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">input</span>, data.type <span class="op">=</span> <span class="st">"fl"</span>,</span>
<span>     exclude.conc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>     log.y <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>     legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>     basesize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     legend.ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     lwd <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/wW5fj0X.png" alt="\label{fig:raw-data-plot} Raw data plot.
Conditions can be selected or deselected using the `names = c('grp1', 'grp2')` argument or `exclude.nm = c('grp3', 'grp4')` argument, respectively. Similarly, concentrations can be (de-selected) via the `conc` and `exclude.conc` arguments. To plot individual samples instead of grouping replicates, add `mean = FALSE`. See `?plot.grodata` for further options." width="80%"><p class="caption">
Raw data plot. Conditions can be selected or deselected using the
<code>names = c('grp1', 'grp2')</code> argument or
<code>exclude.nm = c('grp3', 'grp4')</code> argument, respectively.
Similarly, concentrations can be (de-selected) via the <code>conc</code>
and <code>exclude.conc</code> arguments. To plot individual samples
instead of grouping replicates, add <code>mean = FALSE</code>. See
<code><a href="../reference/plot.grodata.html">?plot.grodata</a></code> for further options.
</p>
</div>
</div>
<div class="section level3">
<h3 id="run-workflow">Run Workflow<a class="anchor" aria-label="anchor" href="#run-workflow"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flFitRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fl.workflow.html">fl.workflow</a></span><span class="op">(</span>grodata <span class="op">=</span> <span class="va">input</span>, </span>
<span>    fit.opt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s"</span>, <span class="st">"l"</span><span class="op">)</span>,</span>
<span>    x_type <span class="op">=</span> <span class="st">"time"</span>, </span>
<span>    norm_fl <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    ec50 <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    dr.parameter <span class="op">=</span> <span class="st">"dY.spline"</span>,</span>
<span>    suppress.messages <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>    export.res <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># Prevent creating TXT table and RData files with results</span></span>
<span>    parallelize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># Use only one available CPU core</span></span></code></pre></div>
<p>If option <span style="color: red;"><code>export.res</code></span> is
set to <span style="color: red;"><code>TRUE</code></span>, tab-delimited
.txt files summarizing the computation results are created
automatically, as well as the <code>flFitRes</code> object (an object of
class <code>flFitRes</code>) as .RData file. This object (or the .RData
file) contains all raw data, fitting options, and computational results.
Figure @ref(fig:flFitRes-container) shows the structure of the generated
<code>flFitRes</code> object. In RStudio, <code>View(flFitRes)</code>
allows interactive inspection of the data container.</p>
<p>If you want to create a report summarizing all computational results
including a graphical representation of every fit, provide the desired
output format(s) as <span style="color: red;"><code>report = 'pdf'</code></span>, <span style="color: red;"><code>report = 'html'</code></span>, or <span style="color: red;"><code>report = c('pdf', 'html')</code></span>. The
advantage of having the report in HTML format is that every figure can
be exported as (editable) PDF file.</p>
<p><em><span style="color: orange;">In the spirit of good scientific
practice (data transparency), I would encourage anyone using QurvE to
attach the .RData file and generated reports to their
publication.</span></em></p>
<p>Arguments that are commonly modified:</p>
<table class="table"><tbody>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>fit.opt</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Which growth fitting methods to perform; a string containing
<code>'l'</code> for linear fits or <code>'s'</code> for spline fits.
Both fit types can be selected as a vector of strings:
<code>c('l', 's').</code>
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>x_type</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Data type used as independent variable. Either <code>'growth'</code> or
<code>'time'</code>.
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>norm_fl</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Use normalized fluorescence values for curve fitting (if
<code>x_type = 'time'</code>).
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>log.y.lin</code></span> <span style="color: red;"><code>log.y.spline</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Should Ln(y/y0) be applied to the fluorescence data for the respective
fits?
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>biphasic</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Extract parameters for two different phases (as observed with, e.g.,
diauxic shifts)
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>interactive</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Controls interactive mode. If <code>TRUE</code>, each fit is visualized
in the Plots pane and the user can adjust fitting parameters and confirm
the reliability of each fit per sample
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>nboot.fl</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Number of bootstrap samples used for nonparametric curve fitting. See
<code><a href="../reference/flBootSpline.html">?flBootSpline</a></code> for details.
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>dr.method</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
Define the method used to perform a dose-responde analysis: smooth
spline fit (<code>'spline'</code>) or model fitting
(<code>'model'</code>, the default). See section 4
</td>
</tr>
<tr>
<td style="text-align:left;width: 3cm; vertical-align: top !important; ">
<span style="color: red;"><code>dr.parameter</code></span>
</td>
<td style="text-align:left;vertical-align: top !important; ">
The response parameter in the output table to be used for creating a
dose response curve. See <code><a href="../reference/fl.drFit.html">?fl.drFit</a></code> for further details.
</td>
</tr>
</tbody></table>
<p><br>
Please consult <code><a href="../reference/fl.workflow.html">?fl.workflow</a></code> for further arguments to
customize the workflow.<br><br></p>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/QJ2KAT8.jpg" alt="\label{fig:flFitRes-container} Internal structure of a `flFitRes`object generated by `growth.workflow()`." width="90%"><p class="caption">
Internal structure of a <code>flFitRes</code>object generated by
<code><a href="../reference/growth.workflow.html">growth.workflow()</a></code>.
</p>
</div>
<div style="page-break-after: always;"></div>
</div>
<div class="section level3">
<h3 id="tabular-results">Tabular results<a class="anchor" aria-label="anchor" href="#tabular-results"></a>
</h3>
<p>A <code>flFitRes</code> object contains two tables summarizing the
computational results: - <code>flFitRes$flFit$flable</code> lists all
calculated curve parameters for every sample and fit -
<code>flFitRes$drFit$drTable</code> contains the results of the
dose-response analysis</p>
<p>Additionally, the dedicated functions
<code><a href="../reference/table_group_fluorescence_linear.html">table_group_fluorescence_linear()</a></code> and
<code><a href="../reference/table_group_fluorescence_spline.html">table_group_fluorescence_spline()</a></code> allow the generation of
grouped results tables for the two fit types with averages and standard
deviations. The column headers in the resulting data frames are
formatted with HTML for visualization in shiny and with
<code><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">DT::datatable()</a></code>.</p>
<p>A summary of results for each individual fit can be obtained by
applying the generic function <code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code> to any fit object
within <code>flFitRes</code>.</p>
</div>
<div class="section level3">
<h3 id="visualize-results">Visualize results<a class="anchor" aria-label="anchor" href="#visualize-results"></a>
</h3>
<p>Several generic <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> allow plotting of results by
merely accessing list items within the <code>flFitRes</code> object
structure (Figure @ref(fig:flFitRes-container)).</p>
<div class="section level4">
<h4 id="inspect-individual-fits">Inspect individual fits<a class="anchor" aria-label="anchor" href="#inspect-individual-fits"></a>
</h4>
<p>It is important to verify the accuracy of the fits that have been
applied before attempting to interpret any results (if the workflow is
not run with <code>interactive = TRUE</code>. This is especially
important when analyzing fluorescence data, as the curve shapes and the
level of noise can vary significantly depending on the specific
experiment and the equipment used for cultivation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedLinear</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, cex.lab <span class="op">=</span> <span class="fl">1.2</span>, cex.axis <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedLinear</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, cex.lab <span class="op">=</span> <span class="fl">1.2</span>, cex.axis <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedLinear</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span>, cex.lab <span class="op">=</span> <span class="fl">1.2</span>, cex.axis <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/919wCo0.png" alt="\label{fig:plot-linear} Linear fit plots to validate the applied fits. For details about this function, run `?plot.gcFitLinear`." width="70%"><img src="https://i.imgur.com/Q90TdEm.png" alt="\label{fig:plot-linear} Linear fit plots to validate the applied fits. For details about this function, run `?plot.gcFitLinear`." width="70%"><img src="https://i.imgur.com/hRQoiOp.png" alt="\label{fig:plot-linear} Linear fit plots to validate the applied fits. For details about this function, run `?plot.gcFitLinear`." width="70%"><p class="caption">
Linear fit plots to validate the applied fits. For details about this
function, run <code><a href="../reference/plot.gcFitLinear.html">?plot.gcFitLinear</a></code>.
</p>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedSpline</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, basesize <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedSpline</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, basesize <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedSpline</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span>, basesize <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/B6bqitM.png" alt="\label{fig:plot-spline} Spline fit plots to validate the applied fits. For details about this function, run `?plot.gcFitLinear`." width="70%"><img src="https://i.imgur.com/HbKpqp9.png" alt="\label{fig:plot-spline} Spline fit plots to validate the applied fits. For details about this function, run `?plot.gcFitLinear`." width="70%"><img src="https://i.imgur.com/vsuSSrq.png" alt="\label{fig:plot-spline} Spline fit plots to validate the applied fits. For details about this function, run `?plot.gcFitLinear`." width="70%"><p class="caption">
Spline fit plots to validate the applied fits. For details about this
function, run <code><a href="../reference/plot.gcFitLinear.html">?plot.gcFitLinear</a></code>.
</p>
</div>
<p>Normalization of fluorescence reads typically introduces additional
noise. While the default smoothing parameter <code>smooth.fl</code> was
suitable to produce good-quality representations of the curves via
nonparametric fits, the linear fits either failed or produced regression
windows that were too small. In order to obtain linear regression
results that accurately represent the linear-increase section of the
curves, we have to decrease the R2 threshold and manually increase the
size of the sliding window (by default calculated automatically for each
sample). These new settings need to be applied to all samples, so we
re-run the entire workflow with adjusted parameters:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flFitRes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fl.workflow.html">fl.workflow</a></span><span class="op">(</span>grodata <span class="op">=</span> <span class="va">input</span>,</span>
<span>                          fit.opt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"s"</span>, <span class="st">"l"</span><span class="op">)</span>,</span>
<span>                          x_type <span class="op">=</span> <span class="st">"time"</span>,</span>
<span>                          norm_fl <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          lin.R2 <span class="op">=</span> <span class="fl">0.95</span>, <span class="co"># Decreased R2 threshold</span></span>
<span>                          lin.h <span class="op">=</span> <span class="fl">20</span>, <span class="co"># Manually defined sliding window size</span></span>
<span>                          ec50 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          dr.parameter <span class="op">=</span> <span class="st">"dY.spline"</span>,</span>
<span>                          suppress.messages <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          export.res <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          parallelize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedLinear</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, cex.lab <span class="op">=</span> <span class="fl">1.2</span>, cex.axis <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedLinear</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, cex.lab <span class="op">=</span> <span class="fl">1.2</span>, cex.axis <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">flFit</span><span class="op">$</span><span class="va">flFittedLinear</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span>, cex.lab <span class="op">=</span> <span class="fl">1.2</span>, cex.axis <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/b2eUR7y.png" alt="\label{fig:plot-linear2} Linear fit plots to validate the linear regressions after re-running the workflow with adjusted parameters." width="70%"><img src="https://i.imgur.com/jRNsTH1.png" alt="\label{fig:plot-linear2} Linear fit plots to validate the linear regressions after re-running the workflow with adjusted parameters." width="70%"><img src="https://i.imgur.com/lvz5Gyn.png" alt="\label{fig:plot-linear2} Linear fit plots to validate the linear regressions after re-running the workflow with adjusted parameters." width="70%"><p class="caption">
Linear fit plots to validate the linear regressions after re-running the
workflow with adjusted parameters.
</p>
</div>
</div>
<div class="section level4">
<h4 id="grouped-spline-fits">Grouped spline fits<a class="anchor" aria-label="anchor" href="#grouped-spline-fits"></a>
</h4>
<p>Applying <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> to the <code>flFitRes</code> object
produces a figure of all spline fits performed as well as the first
derivative (slope) over time. The generic function calls
<code><a href="../reference/plot.flFitRes.html">plot.flFitRes()</a></code> with <code>data.type = 'spline'</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span>,</span>
<span>     data.type <span class="op">=</span> <span class="st">"spline"</span>,</span>
<span>     deriv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>     legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>     legend.ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>     n.ybreaks <span class="op">=</span> <span class="fl">10</span>,</span>
<span>     basesize<span class="op">=</span><span class="fl">10</span>,</span>
<span>     lwd <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/BptmI2k.png" alt="\label{fig:group-spline-plot} Combined plot of all spline fits performed.
In addition to the options available with `data.type = 'raw'`, further arguments can be defined that control the appearance of the secondary panel showing the slope over time. See `?plot.flFitRes` for all options." width="80%"><p class="caption">
Combined plot of all spline fits performed. In addition to the options
available with <code>data.type = 'raw'</code>, further arguments can be
defined that control the appearance of the secondary panel showing the
slope over time. See <code><a href="../reference/plot.flFitRes.html">?plot.flFitRes</a></code> for all options.
</p>
</div>
<p>By arranging the individual samples in a grid, we can create a visual
representation similar to a heat map that illustrates the values of a
chosen parameter. This can be a helpful way to gain insights and
understand trends within the data.:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot.grid.html">plot.grid</a></span><span class="op">(</span><span class="va">flFitRes</span>,</span>
<span>          param <span class="op">=</span> <span class="st">"max_slope.spline"</span>,</span>
<span>          pal <span class="op">=</span> <span class="st">"Mint"</span>,</span>
<span>          log.y <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>          basesize <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/hx1RGzp.png" alt="\label{fig:plot} Plot grid of all spline fits performed with one concentration per row. See `?plot.grid` for available options." width="90%"><p class="caption">
Plot grid of all spline fits performed with one concentration per row.
See <code><a href="../reference/plot.grid.html">?plot.grid</a></code> for available options.
</p>
</div>
<div style="page-break-after: always;"></div>
</div>
<div class="section level4">
<h4 id="compare-growth-parameters">Compare growth parameters<a class="anchor" aria-label="anchor" href="#compare-growth-parameters"></a>
</h4>
<p>The function <code><a href="../reference/plot.parameter.html">plot.parameter()</a></code> works also on
<code>flFitRes</code> objects to compare computed curve parameters:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Parameters obtained from linear regression</span></span>
<span><span class="fu"><a href="../reference/plot.parameter.html">plot.parameter</a></span><span class="op">(</span><span class="va">flFitRes</span>, param <span class="op">=</span> <span class="st">"max_slope.linfit"</span>, basesize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot.parameter.html">plot.parameter</a></span><span class="op">(</span><span class="va">flFitRes</span>, param <span class="op">=</span> <span class="st">"dY.linfit"</span>, basesize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Parameters obtained from nonparametric fits</span></span>
<span><span class="fu"><a href="../reference/plot.parameter.html">plot.parameter</a></span><span class="op">(</span><span class="va">flFitRes</span>, param <span class="op">=</span> <span class="st">"max_slope.spline"</span>, basesize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot.parameter.html">plot.parameter</a></span><span class="op">(</span><span class="va">flFitRes</span>, param <span class="op">=</span> <span class="st">"dY.spline"</span>, basesize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/x8iAear.png" alt="\label{fig:plot-parameter} Parameter plots. If `mean = TRUE`, the results of replicates are combined and shown as their mean ± 95\% confidence interval. As with the functions for combining different growth curves, the arguments `name`, `exclude.nm`, `conc` and `exclude.conc` allow (de)selection of specific samples or conditions. Since we applied growth models to log-transformed data, calling 'dY.orig.model' or 'A.orig.model' instead of 'dY.model' or 'A.model' provides the respective values on the original scale. For linear and spline fits, this is done automatically. For details about this function, run `?plot.parameter`." width="47%"><img src="https://i.imgur.com/tm4rM4h.png" alt="\label{fig:plot-parameter} Parameter plots. If `mean = TRUE`, the results of replicates are combined and shown as their mean ± 95\% confidence interval. As with the functions for combining different growth curves, the arguments `name`, `exclude.nm`, `conc` and `exclude.conc` allow (de)selection of specific samples or conditions. Since we applied growth models to log-transformed data, calling 'dY.orig.model' or 'A.orig.model' instead of 'dY.model' or 'A.model' provides the respective values on the original scale. For linear and spline fits, this is done automatically. For details about this function, run `?plot.parameter`." width="47%"><img src="https://i.imgur.com/DyukJ72.png" alt="\label{fig:plot-parameter} Parameter plots. If `mean = TRUE`, the results of replicates are combined and shown as their mean ± 95\% confidence interval. As with the functions for combining different growth curves, the arguments `name`, `exclude.nm`, `conc` and `exclude.conc` allow (de)selection of specific samples or conditions. Since we applied growth models to log-transformed data, calling 'dY.orig.model' or 'A.orig.model' instead of 'dY.model' or 'A.model' provides the respective values on the original scale. For linear and spline fits, this is done automatically. For details about this function, run `?plot.parameter`." width="47%"><img src="https://i.imgur.com/59Km0tg.png" alt="\label{fig:plot-parameter} Parameter plots. If `mean = TRUE`, the results of replicates are combined and shown as their mean ± 95\% confidence interval. As with the functions for combining different growth curves, the arguments `name`, `exclude.nm`, `conc` and `exclude.conc` allow (de)selection of specific samples or conditions. Since we applied growth models to log-transformed data, calling 'dY.orig.model' or 'A.orig.model' instead of 'dY.model' or 'A.model' provides the respective values on the original scale. For linear and spline fits, this is done automatically. For details about this function, run `?plot.parameter`." width="47%"><p class="caption">
Parameter plots. If <code>mean = TRUE</code>, the results of replicates
are combined and shown as their mean ± 95% confidence interval. As with
the functions for combining different growth curves, the arguments
<code>name</code>, <code>exclude.nm</code>, <code>conc</code> and
<code>exclude.conc</code> allow (de)selection of specific samples or
conditions. Since we applied growth models to log-transformed data,
calling ‘dY.orig.model’ or ‘A.orig.model’ instead of ‘dY.model’ or
‘A.model’ provides the respective values on the original scale. For
linear and spline fits, this is done automatically. For details about
this function, run <code><a href="../reference/plot.parameter.html">?plot.parameter</a></code>.
</p>
</div>
</div>
<div class="section level4">
<h4 id="dose-response-analysis">Dose-response analysis<a class="anchor" aria-label="anchor" href="#dose-response-analysis"></a>
</h4>
<p>The results of the dose-response analysis can be visualized by
calling <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the <code>drFit</code> object that is
stored within <code>flFitRes</code>. This action calls
<code><a href="../reference/plot.drFit.html">plot.drFit()</a></code> which in turn runs
<code><a href="../reference/plot.drFitSpline.html">plot.drFitSpline()</a></code> or <code><a href="../reference/plot.drFitModel.html">plot.drFitModel()</a></code>
(depending on the choice of in the workflow) on every condition for
which a dose-response analysis has been performed. Alternatively, you
can call <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> on the list elements in
<code>grofit$drFit$drFittedModels</code> or
<code>grofit$drFit$drFittedSplines</code>, respectively.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">flFitRes</span><span class="op">$</span><span class="va">drFit</span>, cex.point <span class="op">=</span> <span class="fl">1</span>, cex.lab <span class="op">=</span> <span class="fl">1.1</span>, cex.axis <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="https://i.imgur.com/Q5Onz3F.png" alt="\label{fig:plot-drFit} Dose response analysis - model fits. For details about this function, run `?plot.drFit`." width="70%"><img src="https://i.imgur.com/5mZDIPQ.png" alt="\label{fig:plot-drFit} Dose response analysis - model fits. For details about this function, run `?plot.drFit`." width="70%"><img src="https://i.imgur.com/LfQ7C5o.png" alt="\label{fig:plot-drFit} Dose response analysis - model fits. For details about this function, run `?plot.drFit`." width="70%"><p class="caption">
Dose response analysis - model fits. For details about this function,
run <code><a href="../reference/plot.drFit.html">?plot.drFit</a></code>.
</p>
</div>
<div style="page-break-after: always;"></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-meyer2019" class="csl-entry">
Meyer A.J., Segall-Shapiro T.H., Glassey E., Zhang J. &amp; Voigt C.A.
(2019). Escherichia coli <span>“</span>marionette<span>”</span> strains
with 12 highly optimized small-molecule sensors. Nature Chemical Biology
15 (2): 196–204. <a href="https://doi.org/10.1038/s41589-018-0168-3" class="external-link">https://doi.org/10.1038/s41589-018-0168-3</a>.
</div>
<div id="ref-nikel2022" class="csl-entry">
Nikel P.I., Benedetti I., Wirth N.T., Lorenzo V. de &amp; Calles B.
(2022). Standardization of regulatory nodes for engineering heterologous
gene expression: a feasibility study. Microbial biotechnology 15 (8):
2250–2265. <a href="https://doi.org/10.1111/1751-7915.14063" class="external-link">https://doi.org/10.1111/1751-7915.14063</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nicolas T. Wirth, Jonathan Funk.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
